stdlib/public/runtime/SwiftObject.mm

```diff
@@ -390,36 +395,39 @@ - (BOOL)isEqual:(id)other {
   if (self == other) {
     return YES;
   }
-  // Both objects must be `SwiftObject` in ObjC
-  if (object_getClass(other) != object_getClass((id)self)) {
-    return NO;
-  }
-
-  // TODO: Bincompat check -- return NO if this is an old executable
 
   // Get Swift type for self and other
   auto selfMetadata = _swift_getClassOfAllocated(self);
   auto otherMetadata = _swift_getClassOfAllocated(other);
 
-  // Find common parent Swift class
+  // Find common parent class
   const ClassMetadata * parentMetadata = NULL;
-  // Collect parents of self
-  std::unordered_set<const ClassMetadata *> selfParents;
-  for (auto m = selfMetadata; m != NULL; m = m->Superclass) {
-    selfParents.emplace(m);
-  }
-  // Find first parent of other that is also a parent of self
-  for (auto m = otherMetadata; m != NULL; m = m->Superclass) {
-    if (selfParents.find(m) != selfParents.end()) {
-      parentMetadata = m;
-      break;
+  if (selfMetadata == otherMetadata) {
```

Comment:
In theory, it should be good enough to do `swift_conformsToProtocol` on `selfMetadata`, use `findConformingSuperclass` to find the type that actually contains the conformance, then see if `otherMetadata` is that type or a subtype. We don't allow subclasses to provide a different conformance than a superclass, so this would find the only allowable common conformance, if there is one. It would also avoid traversing the entire superclass chain when there is a conformance, and would cache the negative result to avoid traversing the superclass chain more than once when there isn't.

However, it's still possible to have multiple conformances in a hierarchy, even though we don't allow it, so there are situations where that approach would provide a different answer. So I'm not sure if it's actually a good idea. Just putting it out there as food for thought.

Reply:
I tried your idea.  Let me know if this looks right to you.  In particular, thanks for pointing out that some of these lookups get cached.

Reply:
Looks good! I should also note that we do have a warning when we find multiple conformances in a hierarchy, so there's at least some feedback when that goes wrong: https://github.com/apple/swift/blob/main/stdlib/public/runtime/ProtocolConformance.cpp#L1119